wp/lib/bap_wp/src/equality.ml

```diff
@@ -0,0 +1,354 @@
+open !Core_kernel
+open Bap.Std
+open Graphlib.Std
+open Base__Option.Monad_infix
+
+module Ir = Bap.Std.Graphs.Ir
+module OcamlGraph = Graphlib.To_ocamlgraph (Ir)
+module BFS = Graph.Traverse.Bfs (OcamlGraph)
+
+(* Replaces the real variables with indices with only their base *)
+let strip_indices (e : Exp.t) : Exp.t =
+  let visitor =
+    (object
+      inherit Exp.mapper as super
+      method! map_let var ~exp:exp1 ~body:exp2 =
+        Let (Var.base var, super#map_exp exp1, super#map_exp exp2)
+      method! map_var v = Var (Var.base v)
+    end) in
+  visitor#map_exp e
+
+(* Perform variable lookup in map. BOTH variables must be virtual.
+   If they are not, then fail and return None.*)
+let map_var (vmap : Var.t Var.Map.t)
+    (v1 : Var.t) (v2 : Var.t) : (Var.t Var.Map.t * Var.t) option  =
+  match Var.Map.find vmap v1 with
+  (* if not found in map, add to map *)
+  | None ->
+    (* if is virtual, do mapping *)
+    if Var.is_virtual v1 && Var.is_virtual v2 then
+      Some (Var.Map.add_exn vmap ~key:v1 ~data:v2, v2)
+      (* if is physical register, just return original variable *)
+    else Some (vmap, v1)
+  (* if found in map, return what is found *)
+  | Some v_found -> Some (vmap, v_found)
+
+(* Maps all virtual variables from e1 to their analagous variable within vmap;
+   if a variable not in vmap, is added to vmap from e2.
+   Returns None if cannot map variables because subs do not match in structure *)
+let rec sub_exp (vmap : Var.t Var.Map.t)
```

Comment:
I'd separate this out into two functions. One that performs substitution and one that compare_exp